exchanges/stream/websocket.go

```diff
@@ -593,24 +573,19 @@ func (w *Websocket) trafficMonitor() {
 					default:
 					}
 				}
-				w.setConnectedStatus(true)
+				w.setState(connected)
 				trafficTimer.Reset(w.trafficTimeout)
 			case <-trafficTimer.C: // Falls through when timer runs out
 				if w.verbose {
-					log.Warnf(log.WebsocketMgr,
-						"%v websocket: has not received a traffic alert in %v. Reconnecting",
-						w.exchangeName,
-						w.trafficTimeout)
+					log.Warnf(log.WebsocketMgr, "%v websocket: has not received a traffic alert in %v. Reconnecting", w.exchangeName, w.trafficTimeout)
 				}
 				trafficTimer.Stop()
 				w.setTrafficMonitorRunning(false)
 				w.Wg.Done() // without this the w.Shutdown() call below will deadlock
 				if !w.IsConnecting() && w.IsConnected() {
```

Comment:
Right, yep.
I had to think about this for a second.
We don't have any traffic, but we're trying to reconnect already, so we shouldn't do anything.
However I think this implies a bigger change to handle Connecting properly in this context.

And now I'm looking at it my eye is caught by the "Routine pausing mechanism" briefly

Reply:
Fixed cff39913
> Websocket: Fix and simplify traffic monitor
> 
> trafficMonitor had a check throttle at the end of the for loop to stop it just gobbling the (blocking) trafficAlert channel non-stop.
> That makes sense, except that nothing is sent to the trafficAlert channel if there's no listener.
> So that means that it's out by one second on the trafficAlert, because any traffic received during the pause is doesn't try to send a traffic alert.
> 
> The unstopped timer is deliberately leaked for later GC when shutdown.
> It won't delay/block anything, and it's a trivial memory leak during an infrequent event.
> 
> Deliberately Choosing to recreate the timer each time instead of using Stop, drain and reset

Also accidentally 1c7441c

> Websocket: Use atomics instead of mutex
> 
> This was spurred by looking at the setState call in trafficMonitor and
> the effect on blocking and efficiency.
> With the new atomic types in Go 1.19, and the small types in use here,
> atomics should be safe for our usage. bools should be truly atomic,
> and uint32 is atomic when the accepted value range is less than one byte/uint8 since
> that can be written atomicly by concurrent processors.
> Maybe that's not even a factor any more, however we don't even have to worry enough to check.
> 
> 

Tomorrow I'll check if the existing tests of trafficMonitor are sufficient, since this was a heftier rewrite than intended.

Reply:
Tests extended to cover all trafficMonitor behaviours.
Also caught several bugs in doing so. Embarrassingly large ones. Go tests! 🎉 ( Excuse the pun )

Rebased on master to get latest linter versions on this PR.